### PR TITLE
Refactor `getWidgetLayout`, `combineWidgets` and widget components to rely on grid components (#2671)

### PR DESCRIPTION
**Pull request by johnPhillips**
## Summary

Addresses issue #2671

## Relevant technical choices

Tried to follow the AC and IB as closely as I could, as this is my first foray into this part of the codebase.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/googlesitekit/widgets/components/WidgetWrapper.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	if ( gridColumnWidth === 0 ) {
+		// Widget with columnWidth of 0 should be hidden
+		return (
+			<div className={ HIDDEN_CLASS }>
+				{ children }
+			</div>
+		);
+	} else if ( gridColumnWidth >= 6 ) {
+		return (
+			<Cell size={ gridColumnWidth }>
```
Comment (johnPhillips) [inline assets/js/googlesitekit/widgets/components/WidgetWrapper.js:40]: I left a [comment on the issue](https://github.com/google/site-kit-wp/issues/2671#issuecomment-792789473) regarding this. I have implemented this according to the IB / AC, but I think we may introduce a visual regression here. 
Happy to do as the reviewer recommends.

File: assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js [hunk 2]
```diff
@@ -53,26 +52,31 @@ export default function WidgetAreaRenderer( { slug, totalAreas } ) {
...
-			key={ widget.slug }
-			slug={ widget.slug }
-		/>
-	) );
+	const widgetsOutput = widgets.map( ( widget, i ) => {
+		return (
+			<WidgetWrapper
+				key={ `${ widget.slug }-wrapper` }
+				gridColumnWidth={ gridColumnWidths[ i ] }
+			>
```
Comment (eugene-manuilov) [inline assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js:68]: Let's use a shorter notation:

```javascript
const widgetsOutput = widgets.map( ( widget, i ) => (
    <WidgetWrapper
        key={ `${ widget.slug }-wrapper` }
        gridColumnWidth={ gridColumnWidths[ i ] }
    >
        ...
    </WidgetWrapper>
) );
```

File: assets/js/googlesitekit/widgets/components/WidgetRenderer.js [hunk 3]
```diff
@@ -34,10 +34,11 @@ import { STORE_NAME } from '../datastore/constants';
 import BaseWidget from './Widget';
 import { getWidgetComponentProps } from '../util';
 import { HIDDEN_CLASS } from '../util/constants';
+// import { Cell } from '../../../material-components';
```
Comment (eugene-manuilov) [inline assets/js/googlesitekit/widgets/components/WidgetRenderer.js:37]:
```suggestion
```

File: assets/js/googlesitekit/widgets/components/WidgetWrapper.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { HIDDEN_CLASS } from '../util/constants';
+import { Cell } from '../../../material-components';
+
+const WidgetWrapper = ( { gridColumnWidth, children } ) => {
+	switch ( true ) {
+		case 0 === gridColumnWidth:
```
Comment (eugene-manuilov) [inline assets/js/googlesitekit/widgets/components/WidgetWrapper.js:32]: @johnPhillips let's better use `if` conditions instead of this fancy switch:

```javascript
if ( 0 === gridColumnWidth ) {
    return ...;
}

if ( gridColumnWidth === 3 || gridColumnWidth === 4 ) {
    return ...;
}

if ( gridColumnWidth === 6 || gridColumnWidth === 8 ) {
    return ...;
}

...
```

File: assets/js/googlesitekit/widgets/components/WidgetWrapper.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { HIDDEN_CLASS } from '../util/constants';
+import { Cell } from '../../../material-components';
+
+const WidgetWrapper = ( { gridColumnWidth, children } ) => {
```
Comment (eugene-manuilov) [inline assets/js/googlesitekit/widgets/components/WidgetWrapper.js:30]: @johnPhillips how about other sizes? What if the `gridColumnWidth` equals to 5 or 7? It will return undefined and the widget won't be displayed at all, right?

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @johnPhillips! It's a good start, however, not everything has been implemented yet. See the following statement from the IB that still needs to be addressed:

> Remove the WIDTH_GRID_CLASS_MAP constant and all of it's uses in:
> 
> ...
> assets/js/googlesitekit/widgets/util/constants.js:
> ...

File: assets/js/googlesitekit/widgets/components/WidgetWrapper.js (hunk 5 above)
Comment (johnPhillips) [reply assets/js/googlesitekit/widgets/components/WidgetWrapper.js:30]: Yeah, true. Looking at the code and tests I'm not sure 5 or 7 could happen, but I will handle this just in case.

Comment (johnPhillips) [reply assets/js/googlesitekit/widgets/components/WidgetWrapper.js:30]: Note that I tried to keep the output (in terms of CSS classes in the snapshots and unit tests etc.) the same after this refactor as it was before, and there doesn't seem to be any pre-existing examples of 5 or 7 that I can see. I added these to be handled like `12` as I'm unsure how the mobile / tablet sizing would go in this scenario.

**Review by eugene-manuilov — DISMISSED** [review]
LGTM

File: assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js [hunk 6]
```diff
@@ -53,25 +52,28 @@ export default function WidgetAreaRenderer( { slug, totalAreas } ) {
...
-		<WidgetRenderer
-			gridClassName={ classnames( gridClassNames[ i ] ) }
-			OverrideComponent={ overrideComponents[ i ] ? () => {
-				const { Component, metadata } = overrideComponents[ i ];
-				return <Component { ...metadata } />;
-			} : undefined }
-			key={ widget.slug }
-			slug={ widget.slug }
-		/>
+		<WidgetWrapper
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js:64]: Maybe let's use a more specific name here? How about `WidgetCellWrapper`? While technically it can also wrap the widget in a div that hides it, the primary purpose is to wrap it in a grid cell.

File: assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js [hunk 7]
```diff
@@ -53,25 +52,28 @@ export default function WidgetAreaRenderer( { slug, totalAreas } ) {
...
+			key={ `${ widget.slug }-wrapper` }
+			gridColumnWidth={ gridColumnWidths[ i ] }
+		>
+			<WidgetRenderer
+				OverrideComponent={ overrideComponents[ i ] ? () => {
+					const { Component, metadata } = overrideComponents[ i ];
+					return <Component { ...metadata } />;
+				} : undefined }
+				key={ widget.slug }
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js:73]: This prop shouldn't be necessary anymore now that it's not the top component in the loop.

File: assets/js/googlesitekit/widgets/components/WidgetWrapper.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+		);
+	}
+
+	if ( gridColumnWidth === 6 || gridColumnWidth === 8 ) {
+		return (
+			<Cell lgSize={ gridColumnWidth } mdSize={ 8 }>
+				{ children }
+			</Cell>
+		);
+	}
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/components/WidgetWrapper.js:54]: This doesn't fully match the AC requirements:
1. The extra rule for `6` and `8` shouldn't be needed; in that case it should simply be a single `size`.
2. The rule for `3` and `4` shouldn't be that specific. While the ACs mention that these are realistically the only values that would be used, it states that this should apply to any width lower than 6 (except 0 of course).

So basically, we can remove one condition here completely while making the other apply to anything below 6. Unless there was a specific reason you diverged from the ACs here?

File: assets/js/googlesitekit/widgets/util/get-widget-layout.js [hunk 9]
```diff
@@ -196,7 +202,18 @@ export function getWidgetLayout( widgets, widgetStates ) {
...
+				// Extract the columns on the current row to be resized (we
+				// don't want to include the overflowing widget or previously
+				// calculated rows)
+				let resizeRow = rowColMap[ rowIndex ];
+				// Resize them
+				[ resizeRow, counter ] = resizeColumns( resizeRow, counter );
+				// Insert the resized row back in
+				rowColMap[ rowIndex ] = resizeRow;
+				// Update the columnWidths with the resized row
+				columnWidths = Object.values( rowColMap ).flat();
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/util/get-widget-layout.js:216]: Why was the logic here altered and why a new function `getRowColumnMap` introduced? This should work as before, so why not just get rid of `classNames` and leave the other logic unaltered?

**Review by felixarntz — CHANGES_REQUESTED** [review]
This looks great for the most part, but there are some minor suggestions for improvement and a few questions to clarify.

File: assets/js/googlesitekit/widgets/util/get-widget-layout.js (hunk 9 above)
Comment (johnPhillips) [reply assets/js/googlesitekit/widgets/util/get-widget-layout.js:216]: My source of truth while working on this was the snapshots, the rationale being that if they all passed then I had a reasonable hope that the refactor didn't cause a regression. So I only deviated from the IB / ACs when the snapshot tests failed.

In this case, the resize operation seemed to malfunction and some snapshot tests started failing. For example this test:
```
WidgetAreaRenderer should resize widgets in a row that spans 9 columns to fill the full 12 columns 
(3, 12, 3-6, 6 -> 3, 12, 4-8, 6)
```
The resize operation calculated the 9-column row correctly (`3-6 -> 4-8`), but then continued into the previous rows. So here it adjusted the `3` on the first row to be a `4`, meaning the final output was `4, 12, 4-8, 6`.

I found that if I only passed the offending row, it was resized correctly and the tests passed.

I think it's to do with this change, because we lose the `includes` from the original, which might have been why we didn't see this issue with the previous code:
```js
-	if ( singleWidgetClassNames.includes( 'mdc-layout-grid__cell--span-3-desktop' ) ) {
-		singleWidgetClassNames[ singleWidgetClassNames.indexOf( 'mdc-layout-grid__cell--span-3-desktop' ) ] = 'mdc-layout-grid__cell--span-4-desktop';
+	if ( columnWidths[ i ] === 3 ) {
```

That's my best guess though - I'll be honest, I found this quite a tricky ticket! 😂
I'm very happy to remove this if the regression is not an issue. Or maybe I've misunderstood something along the way here.

File: assets/js/googlesitekit/widgets/components/WidgetWrapper.js (hunk 8 above)
Comment (johnPhillips) [reply assets/js/googlesitekit/widgets/components/WidgetWrapper.js:54]: As I explained below, I tried to use the snapshots as the source of truth throughout. 

> The extra rule for 6 and 8 shouldn't be needed; in that case it should simply be a single size.

In this case I deviated from the ACs because I was seeing diffs like this:
### 
```diff
-  class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-8-tablet"
+  class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6"
```
I was concerned that this might cause a regression. It seems that for a `6` and `8` on desktop, we currently do `8` on tablet. Can you advise?

> The rule for 3 and 4 shouldn't be that specific. While the ACs mention that these are realistically the only values that would be used, it states that this should apply to any width lower than 6 (except 0 of course).

I can certainly change this 👍

Comment (felixarntz) [reply assets/js/googlesitekit/widgets/components/WidgetWrapper.js:54]: @johnPhillips Ah right, thanks, that's a good catch. In that case, it should be adjusted as follows:
* If smaller than 6, use the first condition (4 on medium, 2 on small)
* If smaller than 8, use the second condition (8 on medium)
* Otherwise use the default (no fallback, always full amount of columns)

File: assets/js/googlesitekit/widgets/util/get-widget-layout.js (hunk 9 above)
Comment (felixarntz) [reply assets/js/googlesitekit/widgets/util/get-widget-layout.js:216]: @johnPhillips I found the problem here. It was indeed broken before, but since previously we weren't using the column widths (nor really testing them) it never surfaced.

* The problem is https://github.com/google/site-kit-wp/pull/2927/files#diff-1accd10359d13467fb6f9186730cc4ed6a8b0aac9139453bcd5cb3a6f165e807R187
* In that line we add the column width for the current widget to the list of column widths before we have even fitted it into the current row. This is different from before where the class names were only added to the actual area at the very end (see https://github.com/google/site-kit-wp/pull/2927/files#diff-1accd10359d13467fb6f9186730cc4ed6a8b0aac9139453bcd5cb3a6f165e807L214, there's also a comment on why that was done).
* The problem is basically that due to adding the current column width to the list of widths too early, one more entry is in there than it should. Before this PR, this simply meant the column widths were out of sync with the class names passed to `resizeClasses`. Now that we actually rely on the column widths though, this surfaces as a problem, breaking the expected behavior.
* When reverting to the simpler previous code where it was only calling `resizeClasses` here and adding a `console.log` to the beginning of `resizeClasses`, you will see that it's always one too many numbers that is passed to the function. That becomes more apparent when you add the same `console.log` to the current `develop` codebase: In there, you'll be able to see that it's always one more entry in `columnWidths` than in `classNames`, and the last entry is the one that shouldn't be there.

In other words, TLDR:
* Please remove the new extra logic and function you've added here and rely on calling `resizeClasses` as before.
* Instead, to fix the problem remove the early addition to the `columnWidths` array in https://github.com/google/site-kit-wp/pull/2927/files#diff-1accd10359d13467fb6f9186730cc4ed6a8b0aac9139453bcd5cb3a6f165e807R187 and instead do that in https://github.com/google/site-kit-wp/pull/2927/files#diff-1accd10359d13467fb6f9186730cc4ed6a8b0aac9139453bcd5cb3a6f165e807R231.

File: assets/js/googlesitekit/widgets/components/WidgetWrapper.js (hunk 8 above)
Comment (johnPhillips) [reply assets/js/googlesitekit/widgets/components/WidgetWrapper.js:54]: No problem, thanks!

File: assets/js/googlesitekit/widgets/util/get-widget-layout.js (hunk 9 above)
Comment (johnPhillips) [reply assets/js/googlesitekit/widgets/util/get-widget-layout.js:216]: Ah, well spotted! I think that bug required an extra brain in the end 😂

I can't seem to get the link in your comment to work, but I think you meant [this line](https://github.com/google/site-kit-wp/blob/develop/assets/js/googlesitekit/widgets/util/get-widget-layout.js#L181)?
```js
// Populate initial column width and row index for the widget.
// These may be corrected below in certain situations.
columnWidths.push( WIDTH_GRID_COUNTER_MAP[ width ] ); //  ← 💥
rowIndexes.push( rowIndex );
```

When I remove that, the tests pass, so hopefully that's it. Let me know if I misunderstood.

I've removed the extra logic too 👍

**Review by felixarntz — COMMENTED** [review]
Awesome, this is pretty much good for merge!

Just a few nit-picks, I'll quickly address them.

**Review by felixarntz — APPROVED** [review]
